Improve and update `.hea` example

briangow opened the pull request:
This PR improves the `.hea` examples by adding descriptions for each element from a header line. It also updates the example header lines to show a multi-frequency record. This gives users additional insight into how frames, samples per frame, and total samples can be garnered from the `.hea`. 

I've also added a reference to the detailed header documentation for easy reference.

briangow committed:
update hea example

tompollard commented on spec/header-files.md:
```diff
@@ -21,28 +21,35 @@ A header file contains:
-| Base date (optional) | Start date (`DD/MM/YYYY`). |
+| Field | Description                                                                |
+|:------|:---------------------------------------------------------------------------|
+| Record name | Identifier for the record (letters, digits, underscores only).             |
```
Please could you remove the trailing whitespace on these rows? I don't think it should be there?

briangow replied:
I removed the extra spaces. I think PyCharm was introducing them.

tompollard commented on spec/header-files.md:
````diff
@@ -21,28 +21,35 @@ A header file contains:
 **Example:**
 ```text
-100 2 360 650000 12:00:00 01/01/2000
+12345 3 62.5 625 30/01/1989
````
In the new example, we have lost base time but for no particular reason I can see.  Instead of overwriting the existing sample, I wonder whether it would be better to list multiple examples? 

My preference is to use real-life examples if we can, so that we could link to the real file allowing people to do a deeper exploration.

briangow replied:
I can make sure the example has `base_time` set either way.

tompollard commented:
The confusing thing about the current header structure is that the location/order of fields is inconsistent, depending on whether or not an optional field is provided. I like the idea of adding a bulleted list below the examples, which I think partly addresses this issue.

I think it is much better to use real-life examples if possible. I hadn't made it clear (and we should do this) but the example I provided was from the MIT-BIH Arrhythmia Database if remember correctly (though it seems I may have modified a couple of the fields, which I probably shouldn't have done: https://physionet.org/content/mitdb/1.0.0/100.hea).

briangow commented:
@tompollard , thanks for your thoughts!

> The confusing thing about the current header structure is that the location/order of fields is inconsistent, depending on whether or not an optional field is provided. I like the idea of adding a bulleted list below the examples, which I think partly addresses this issue.

I completely agree, which is why I like the idea of providing the explanation below the examples. It certainly isn't a comprehensive solution but should get people started. 

> I think it is much better to use real-life examples if possible.

I agree but I wasn't able to find good multi-frequency WFDB examples on PN. I personally like showing this capability as it gives insight into some of the fields that get brushed over with simpler examples. I also anticipate that CHoRUS will make use of the multi-frequency design, so there will be real-world example data sooner than later. CHoRUS will probably also create multi-segment records but providing examples of these does get significantly more involved. 

It isn't always clear where to strike a balance with all of this. This is one of the reasons I provided a link to the detailed header documentation which does have additional examples. On the wfdb.io site, do you think we should give examples from multiple records? If so, do we want these examples:
- one sample rate for all channels, not multi-segment (could use the old example you had)
- unique sample rates per channel, not multi-segment (example I give in this PR, until we can identify a real-world example)
- unique sample rates per channel, multi-segment record (MIMIC-IV Waveforms - example with only 2 segments: https://physionet.org/content/mimic4wdb/0.1.0/waves/p100/p10082591/84050536 )

tompollard commented:
Not directly related to the PR, but another confusing aspect of WFDB is that the format specification is not well separated from the software documentation. This was on my mind again when reviewing https://github.com/wfdb/wfdb.github.io/pull/10.

My preference is to make this website (or a new specification website) the primary location for detailed specification of the format. Then we link out to documentation for the individual software implementations (as we do now at https://wfdb.io/software/).

> I agree but I wasn't able to find good multi-frequency WFDB examples on PN. I personally like showing this capability as it gives insight into some of the fields that get brushed over with simpler examples.

Your new multi-frequency example is useful and if we can't find a real example then using synthetic values is good with me. I just felt like it supplemented the previous example, so it would be better to keep both (and add additional examples as needed, e.g. the MIMIC-IV Waveforms one you mention).

briangow committed:
add multiple header examples

briangow commented:
@tompollard , I've updated this to include both an example from the file you referenced previously and the multi-frequency example. For consistency, I switched the MIT-BIH example that you were using back to the exact settings in the header file. I also mention that the file comes from that database, so people can further explore it if desired. I added explanations for the elements in that example also. 

If we do end up adding a multi-segment example, we'll probably need to create collapsable elements (using `<detail>` tags). I've found this only creates cleanly formatted output if the entire example is contained inside html elements. My opinion is that using these collapsable elements isn't necessary for just the 2 examples we currently have.

tompollard commented on spec/header-files.md:
```diff
@@ -21,28 +21,49 @@ A header file contains:
-**Example:**
+### Examples
+
+#### Basic header (uniform sampling frequency)(record 100 from the [MIT-BIH database](https://www.physionet.org/content/mitdb)):
```
Instead of having two sets of parentheses in the header like this: 

`Basic header (uniform sampling frequency)(record 100 from the MIT-BIH database)`

...it would be cleaner to put the "record 100 from the MIT-BIH database" note below the header.

briangow replied:
I've updated this but I think it looks better with "record 100 from the MIT-BIH database" immediately following the title instead of on the next line.

tompollard commented on spec/header-files.md:
````diff
@@ -21,28 +21,49 @@ A header file contains:
-100 2 360 650000 12:00:00 01/01/2000
+100 2 360 650000
 ```
+- `100`: Record name (must match the filename prefix).
````
As I'm reading this, I'm trying to match the bullet to one of the fields in the table above. I think the bulleted list would be easier to follow if the descriptive text here exactly matched the "field" text above (e.g. instead of `Total samples for each signal.` we would put `Number of samples`). If necessary, we can provide additional information in parentheses.

briangow replied:
Agreed - I've updated this and added clarification when the table name isn't clear / correct on it's own.

tompollard commented on spec/header-files.md:
````diff
@@ -21,28 +21,49 @@ A header file contains:
+```text
+12345 3 62.5 625 12:00:00 30/01/1989
+```
+- `12345`: Record name (must match the filename prefix).
````
Might not make a difference for this particular markdown renderer (I haven't tried building this locally) but it is good practice to have an empty line before a bulleted list.

briangow replied:
Updated.

tompollard commented on spec/header-files.md:
````diff
@@ -67,11 +88,39 @@ Each signal has its own line immediately following the record line (for single-s
+### Examples
+
+#### Basic header (uniform sampling frequency)(record 100 from the [MIT-BIH database](https://www.physionet.org/content/mitdb)):
+```text
````
Similar to bulleted list. It might not make a difference for this particular renderer but it is good practice to have an empty line after a header (e.g. see: https://www.markdownguide.org/basic-syntax/ - "You should also put blank lines before and after a heading for compatibility.")

briangow replied:
I ended up changing these to bold instead of headers. The fourth level header is too small relative to the pages text.  Thanks for the tip though.

tompollard commented on spec/header-files.md:
````diff
@@ -67,11 +88,39 @@ Each signal has its own line immediately following the record line (for single-s
+100.dat 212 200 11 1024 995 -22131 0 MLII
+100.dat 212 200 11 1024 1011 20052 0 V5
+```
+- `100.dat`: Filename of the signal file.
````
add blank line

briangow replied:
Updated.

tompollard commented on spec/header-files.md:
```diff
@@ -21,28 +21,49 @@ A header file contains:
+- `12345`: Record name (must match the filename prefix).
+- `3`: Number of signals in the record.
+- `62.5`: Sampling frequency in Hz.
+- `625`: Total frames for each signal. When `samps_per_frame` =  1 (default), this will be the total samples.
```
The "Total frames for each signal" line is what prompted me to make the comment above about matching field names to the table exactly. In this case, it's unclear to me whether the "Total frames for each signal" matches to the "Number of samples" field, or to a different field.

briangow replied:
Yeah, I like using the table name and then adding clarification as needed in the example descriptions.

tompollard reviewed (approved):
Thanks Brian, sorry more comments than I'd expected to make. I'll approve the PR, but I think it would be good to address some or all of the comments before merging, if possible.

briangow committed:
header example clarification

tompollard commented:
Thanks for taking the time to address these comments!
